kernel/userspace.c

```diff
@@ -466,8 +466,11 @@ static void unref_check(struct z_object *ko, uintptr_t index)
 	sys_bitfield_clear_bit((mem_addr_t)&ko->perms, index);
 
 #ifdef CONFIG_DYNAMIC_OBJECTS
-	struct dyn_obj *dyn =
-			CONTAINER_OF(ko, struct dyn_obj, kobj);
+	void *vko = ko;
+
+	struct dyn_obj *dyn = CONTAINER_OF(vko, struct dyn_obj, kobj);
+	/* TODO: check why this assert hits */
+	/*__ASSERT(IS_PTR_ALIGNED(dyn, struct dyn_obj), "unaligned z_object");*/
```

Comment:
Just to repeat my comment from elsewhere: this points to some suspicious behavior in the kernel dynamic object allocator.  We should probably open a bug to make sure it gets looked at.  Given the architecture, it's not illegal to misalign structs, but we shouldn't be doing it without a good reason.

(If I had to make a wild guess, it was written for i386 and 4-byte-aligns everything, so misses half the time for structs with 64 bit pointers.)

Reply:
Indeed, see here: https://github.com/zephyrproject-rtos/zephyr/issues/41062